[Bug] Concurrency problems may occur when the ApplicationManager registerApplicationInfo

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)


### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-uniffle/issues?q=is%3Aissue) and found no similar issues.


### Describe the bug

```
appAndTime.remove(uuidFromApp);
appAndTime.put(appId, currentTimeMillis);
```
In the process of `registerApplicationInfo` in `ApplicationManager`, we need to remove the uuid and re insert it into the registered appId. This process is not atomic and may cause concurrency problems.
```
int currentAppNum = appAndTimes.size();
if (currentAppNum >= defaultAppNum) {
    ...
}
```
When `AccessQuotaChecker` getting `appAndTimes.size()`, it may be the number of apps before putting appId after removing uuid.


### Affects Version(s)

0.7.0

### Uniffle Server Log Output

[ERROR] 2022-11-29 17:57:39,386 Grpc-199 AccessAppQuotaChecker check - Denied by AccessAppQuotaChecker => User: xxxx, current app num is: 104, default app num is: 80. We will reject this app[uuid=468366284-108712]

### Uniffle Engine Log Output

_No response_

### Uniffle Server Configurations

_No response_

### Uniffle Engine Configurations

_No response_

### Additional context

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

--- a/coordinator/src/main/java/org/apache/uniffle/coordinator/AccessQuotaChecker.java
+++ b/coordinator/src/main/java/org/apache/uniffle/coordinator/AccessQuotaChecker.java
@@ -18,10 +18,8 @@
 package org.apache.uniffle.coordinator;
 
 import java.io.IOException;
-import java.util.Map;
 import java.util.concurrent.atomic.LongAdder;
 
-import com.google.common.collect.Maps;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -52,23 +50,14 @@ public class AccessQuotaChecker extends AbstractAccessChecker {
     final String uuid = hostIp.hashCode() + "-" + COUNTER.sum();
     final String user = accessInfo.getUser();
     // low version client user attribute is an empty string
-    if (!"".equals(user)) {
-      Map<String, Map<String, Long>> currentUserApps = quotaManager.getCurrentUserAndApp();
-      Map<String, Long> appAndTimes = currentUserApps.computeIfAbsent(user, x -> Maps.newConcurrentMap());
-      Integer defaultAppNum = quotaManager.getDefaultUserApps().getOrDefault(user,
-          conf.getInteger(CoordinatorConf.COORDINATOR_QUOTA_DEFAULT_APP_NUM));
-      synchronized (ApplicationManager.class) {
-        int currentAppNum = appAndTimes.size();
-        if (currentAppNum >= defaultAppNum) {
-          String msg = "Denied by AccessQuotaChecker => "
-              + "User: " + user + ", current app num is: " + currentAppNum
-              + ", default app num is: " + defaultAppNum + ". We will reject this app[uuid=" + uuid + "].";
-          LOG.error(msg);
-          CoordinatorMetrics.counterTotalQuotaDeniedRequest.inc();
-          return new AccessCheckResult(false, msg);
-        }
-        appAndTimes.put(uuid, System.currentTimeMillis());
-      }
+    if (!"".equals(user) && quotaManager.checkQuota(user, uuid)) {
+      String msg = "Denied by AccessQuotaChecker => "
+          + "User: " + user + ", current app num is: " + quotaManager.getCurrentUserAndApp().get(user).size()
+          + ", default app num is: " + quotaManager.getDefaultUserApps().get(user)
+          + ". We will reject this app[uuid=" + uuid + "].";
+      LOG.error(msg);
+      CoordinatorMetrics.counterTotalQuotaDeniedRequest.inc();
+      return new AccessCheckResult(false, msg);
     }
     return new AccessCheckResult(true, Constants.COMMON_SUCCESS_MESSAGE, uuid);
   }

--- a/coordinator/src/main/java/org/apache/uniffle/coordinator/AccessQuotaChecker.java
+++ b/coordinator/src/main/java/org/apache/uniffle/coordinator/AccessQuotaChecker.java
@@ -57,16 +57,18 @@ public class AccessQuotaChecker extends AbstractAccessChecker {
       Map<String, Long> appAndTimes = currentUserApps.computeIfAbsent(user, x -> Maps.newConcurrentMap());
       Integer defaultAppNum = quotaManager.getDefaultUserApps().getOrDefault(user,
           conf.getInteger(CoordinatorConf.COORDINATOR_QUOTA_DEFAULT_APP_NUM));
-      int currentAppNum = appAndTimes.size();
-      if (currentAppNum >= defaultAppNum) {
-        String msg = "Denied by AccessClusterLoadChecker => "
-            + "User: " + user + ", current app num is: " + currentAppNum
-            + ", default app num is: " + defaultAppNum + ". We will reject this app[uuid=" + uuid + "].";
-        LOG.error(msg);
-        CoordinatorMetrics.counterTotalQuotaDeniedRequest.inc();
-        return new AccessCheckResult(false, msg);
+      synchronized (ApplicationManager.class) {
+        int currentAppNum = appAndTimes.size();
+        if (currentAppNum >= defaultAppNum) {
+          String msg = "Denied by AccessQuotaChecker => "
+              + "User: " + user + ", current app num is: " + currentAppNum
+              + ", default app num is: " + defaultAppNum + ". We will reject this app[uuid=" + uuid + "].";
+          LOG.error(msg);
+          CoordinatorMetrics.counterTotalQuotaDeniedRequest.inc();
+          return new AccessCheckResult(false, msg);
+        }
+        appAndTimes.put(uuid, System.currentTimeMillis());
       }
-      appAndTimes.put(uuid, System.currentTimeMillis());
     }
     return new AccessCheckResult(true, Constants.COMMON_SUCCESS_MESSAGE, uuid);
   }

--- a/coordinator/src/main/java/org/apache/uniffle/coordinator/ApplicationManager.java
+++ b/coordinator/src/main/java/org/apache/uniffle/coordinator/ApplicationManager.java
@@ -110,8 +110,10 @@ public class ApplicationManager {
     String[] appIdAndUuid = appId.split("_");
     String uuidFromApp = appIdAndUuid[appIdAndUuid.length - 1];
     // if appId created successfully, we need to remove the uuid
-    appAndTime.remove(uuidFromApp);
-    appAndTime.put(appId, currentTimeMillis);
+    synchronized (ApplicationManager.class) {
+      appAndTime.remove(uuidFromApp);
+      appAndTime.put(appId, currentTimeMillis);
+    }
   }
 
   public void refreshAppId(String appId) {

--- a/coordinator/src/main/java/org/apache/uniffle/coordinator/ApplicationManager.java
+++ b/coordinator/src/main/java/org/apache/uniffle/coordinator/ApplicationManager.java
@@ -106,13 +106,10 @@ public class ApplicationManager {
       CoordinatorMetrics.counterTotalAppNum.inc();
       LOG.info("New application is registered: {}", appId);
     }
-    long currentTimeMillis = System.currentTimeMillis();
-    String[] appIdAndUuid = appId.split("_");
-    String uuidFromApp = appIdAndUuid[appIdAndUuid.length - 1];
-    // if appId created successfully, we need to remove the uuid
-    synchronized (ApplicationManager.class) {
-      appAndTime.remove(uuidFromApp);
-      appAndTime.put(appId, currentTimeMillis);
+    if (quotaManager != null) {
+      quotaManager.registerApplicationInfo(appId, appAndTime);
+    } else {
+      appAndTime.put(appId, System.currentTimeMillis());
     }
   }
 

--- a/coordinator/src/main/java/org/apache/uniffle/coordinator/QuotaManager.java
+++ b/coordinator/src/main/java/org/apache/uniffle/coordinator/QuotaManager.java
@@ -49,12 +49,14 @@ public class QuotaManager {
   private final Map<String, Map<String, Long>> currentUserAndApp = Maps.newConcurrentMap();
   private final Map<String, String> appIdToUser = Maps.newConcurrentMap();
   private final String quotaFilePath;
+  private final Integer quotaAppNum;
   private FileSystem hadoopFileSystem;
   private final AtomicLong quotaFileLastModify = new AtomicLong(0L);
   private final Map<String, Integer> defaultUserApps = Maps.newConcurrentMap();
 
   public QuotaManager(CoordinatorConf conf) {
     this.quotaFilePath = conf.get(CoordinatorConf.COORDINATOR_QUOTA_DEFAULT_PATH);
+    this.quotaAppNum = conf.getInteger(CoordinatorConf.COORDINATOR_QUOTA_DEFAULT_APP_NUM);
     if (quotaFilePath == null) {
       LOG.warn("{} is not configured, each user will use the default quota : {}",
           CoordinatorConf.COORDINATOR_QUOTA_DEFAULT_PATH.key(),
@@ -107,6 +109,31 @@ public class QuotaManager {
       }
     } catch (Exception e) {
       LOG.error("Error occur when parsing file {}", quotaFilePath, e);
+    }
+  }
+
+  public boolean checkQuota(String user, String uuid) {
+    Map<String, Long> appAndTimes = currentUserAndApp.computeIfAbsent(user, x -> Maps.newConcurrentMap());
+    Integer defaultAppNum = defaultUserApps.getOrDefault(user, quotaAppNum);
+    synchronized (this) {
+      int currentAppNum = appAndTimes.size();
+      if (currentAppNum >= defaultAppNum) {
+        return true;
+      } else {
+        appAndTimes.put(uuid, System.currentTimeMillis());
+        return false;
+      }
+    }
+  }
+
+  public void registerApplicationInfo(String appId, Map<String, Long> appAndTime) {
+    long currentTimeMillis = System.currentTimeMillis();
+    String[] appIdAndUuid = appId.split("_");
+    String uuidFromApp = appIdAndUuid[appIdAndUuid.length - 1];
+    // if appId created successfully, we need to remove the uuid
+    synchronized (this) {
+      appAndTime.remove(uuidFromApp);
+      appAndTime.put(appId, currentTimeMillis);
     }
   }
 

--- a/coordinator/src/test/java/org/apache/uniffle/coordinator/QuotaManagerTest.java
+++ b/coordinator/src/test/java/org/apache/uniffle/coordinator/QuotaManagerTest.java
@@ -20,6 +20,9 @@ package org.apache.uniffle.coordinator;
 import java.io.File;
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
 
 import com.google.common.collect.Lists;
 import org.apache.hadoop.conf.Configuration;
@@ -33,6 +36,7 @@ import org.junit.jupiter.api.io.TempDir;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 
 /**
  * QuotaManager is a manager for resource restriction.
@@ -99,5 +103,34 @@ public class QuotaManagerTest {
     Thread.sleep(500);
     // it didn't detectUserResource because `org.apache.unifle.coordinator.AccessQuotaChecker` is not configured
     assertNull(applicationManager.getQuotaManager());
+  }
+
+  @Test
+  public void testCheckQuota() throws Exception {
+    final String quotaFile =
+        new Path(remotePath.getAbsolutePath()).getFileSystem(hdfsConf).getName() + "/quotaFile.properties";
+    CoordinatorConf conf = new CoordinatorConf();
+    conf.set(CoordinatorConf.COORDINATOR_QUOTA_DEFAULT_PATH,
+        quotaFile);
+    final ApplicationManager applicationManager = new ApplicationManager(conf);
+    final AtomicInteger uuid = new AtomicInteger();
+    Map<String, Long> uuidAndTime = new ConcurrentHashMap<>();
+    uuidAndTime.put(String.valueOf(uuid.incrementAndGet()), System.currentTimeMillis());
+    uuidAndTime.put(String.valueOf(uuid.incrementAndGet()), System.currentTimeMillis());
+    uuidAndTime.put(String.valueOf(uuid.incrementAndGet()), System.currentTimeMillis());
+    uuidAndTime.put(String.valueOf(uuid.incrementAndGet()), System.currentTimeMillis());
+    final int i1 = uuid.incrementAndGet();
+    uuidAndTime.put(String.valueOf(i1), System.currentTimeMillis());
+    Map<String, Long> appAndTime = applicationManager.getQuotaManager().getCurrentUserAndApp()
+        .computeIfAbsent("user1", x -> uuidAndTime);
+    // This thread may remove the uuid and put the appId in.
+    final Thread registerThread = new Thread(() ->
+        applicationManager.getQuotaManager().registerApplicationInfo("application_test_" + i1, appAndTime));
+    registerThread.start();
+    final boolean icCheck = applicationManager.getQuotaManager()
+        .checkQuota("user1", String.valueOf(i1));
+    registerThread.join();
+    assertTrue(icCheck);
+    assertEquals(applicationManager.getQuotaManager().getCurrentUserAndApp().get("user1").size(), 5);
   }
 }

